Frameworks/Foundation/NSNotificationCenter.mm

```diff
@@ -14,227 +14,289 @@
+        StrongId<NSMutableSet<_NSNotificationReceiver*>> nameReceivers = [self _receiverSetForNotificationName:name createIfMissing:NO];
+        if (!nameReceivers) {
+            return;
         }
 
-        if (!observer->block) {
-            [observer->object performSelector:observer->selector withObject:notification];
-        } else {
-            observer->block(notification);
+        [nameReceivers minusSet:receivers];
+
+        if ([nameReceivers count] == 0) {
+            // This can release nameReceivers; it will be retained for the scope of this method.
+            [_observers removeObjectForKey:name];
         }
-        [observer release];
+    } else {
+        // Slow path: remove these observers from every name table.
+        NSArray<NSString*>* copiedAllKeys = [[_observers allKeys] copy];
+        for (NSString* key in copiedAllKeys) {
+            StrongId<NSMutableSet<_NSNotificationReceiver*>> nameReceivers = [_observers objectForKey:key];
+            [nameReceivers minusSet:receivers];
+            if ([nameReceivers count] == 0) {
+                [_observers removeObjectForKey:name];
+            }
+        }
+        [copiedAllKeys release];
     }
-    IwFree(sendTo);
 }
 
 /**
  @Status Interoperable
 */
-- (void)postNotificationName:(NSString*)name object:(NSObject*)sender userInfo:(NSDictionary*)userInfo {
-    NSNotification* notification = [NSNotification notificationWithName:name object:sender userInfo:userInfo];
+- (void)postNotificationName:(NSString*)name object:(NSObject*)sender {
+    [self postNotificationName:name object:sender userInfo:nil];
+}
 
+/**
+ @Status Interoperable
+*/
+- (void)postNotificationName:(NSString*)name object:(NSObject*)sender userInfo:(NSDictionary*)userInfo {
+    // Since object could be mid-deallocation, this notification must not be leaked into an autorelease pool.
+    StrongId<NSNotification> notification;
+    notification.attach([[NSNotification alloc] initWithName:name object:sender userInfo:userInfo]);
     [self postNotification:notification];
 }
 
 /**
  @Status Interoperable
 */
-- (void)addObserver:(id)observer selector:(SEL)selName name:(NSString*)name object:(id)object {
-    if (name == nil) {
+- (void)postNotification:(NSNotification*)notification {
+    NSString* name = [notification name];
+
+    // This method does not use _receiversForName:(...) above as it treats a nil _notification_ sender as a wildcard.
+    // We want to treat a nil _receiver_ sender as a wildcard.
+    std::vector<StrongId<_NSNotificationReceiver>> validReceivers{ 0 };
+    @synchronized(self) {
+        NSMutableSet<_NSNotificationReceiver*>* receivers = [self _receiverSetForNotificationName:name createIfMissing:NO];
+        if (!receivers) {
+            return;
+        }
+
+        NSObject* sender = [notification object];
+        validReceivers.reserve([receivers count]);
+        for (_NSNotificationReceiver* receiver in receivers) {
```

Comment:
Not in this case. We want strict control over the lifetimes of these receivers, and we don't want to vend them into an autorelease pool. `objectsPassingTest:` will vend them into an autorelease pool.